src/SAWScript/MGU.hs

```diff
@@ -293,21 +295,39 @@ newTypeIndex = do
   TI $ put $ rw { nameGen = nameGen rw + 1 }
   return $ nameGen rw
 
-newType :: TI Type
-newType = TyUnifyVar <$> newTypeIndex
-
+-- Construct a new type variable.
+--
+-- Collect the position that prompted us to make it; for example, if
+-- we're the element type of an empty list we get the position of the
+-- []. We haven't inferred anything, so use the Placeholder position.
+-- This will cause the position of anything more substantive that gets
+-- unified with it to be preferred. If no such thing happens though
+-- this will be the position that gets attached to the quantifier
+-- binding in generalize.
+newType :: Pos -> TI Type
+newType pos = TyUnifyVar (PosInferred InfFresh pos) <$> newTypeIndex
+
+-- Construct a new type variable to use as a placeholder after an
+-- error occurs. For now this is the same as other fresh type
+-- variables, but I've split it out in case we want to distinguish it
+-- in the future.
+newTypeError :: Pos -> TI Type
+newTypeError pos = newType pos
+
+-- Typecheck a pattern and produce fresh type variables as needed.
+-- FUTURE: this function should get renamed.
```

Comment:
A general comment: it would be worth opening issues for any concrete tasks you identify as future work.

Reply:
I can do that but in general it will create a deluge of low-quality issues about minor code things like this and I'm not sure it's desirable. In this particular case almost everything in here should probably get some level of naming reform, including the filename :-|